In [1]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
# Charger les données MNIST
mnist = pd.read_csv("mnist_784.csv")
# Diviser les données en ensembles d'apprentissage et de test
X_train, X_test = train_test_split(mnist, test_size=40000, random_state=42)

# Supprimer la colonne "class" de l'ensemble d'apprentissage
X_train = X_train.drop(columns=["class"])

In [2]:
# Classification sans ACP
kmeans_no_pca = KMeans(n_clusters=9, random_state=42)
labels_no_pca = kmeans_no_pca.fit_predict(X_train)
silhouette_train_no_pca = silhouette_score(X_train, labels_no_pca)
print("Indice de silhouette sans ACP pour l'ensemble d'apprentissage :", silhouette_train_no_pca)


C:\Users\merye\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Indice de silhouette sans ACP pour l'ensemble d'apprentissage : 0.05670380685602751


In [3]:
# Classification avec ACP à 90% de variance cumulée
pca = PCA(n_components=0.9, random_state=42)
X_train_pca = pca.fit_transform(X_train)
kmeans_with_pca = KMeans(n_clusters=9, random_state=42)
labels_with_pca = kmeans_with_pca.fit_predict(X_train_pca)
silhouette_train_with_pca = silhouette_score(X_train_pca, labels_with_pca)
print("Indice de silhouette avec ACP pour l'ensemble d'apprentissage :", silhouette_train_with_pca)

C:\Users\merye\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Indice de silhouette avec ACP pour l'ensemble d'apprentissage : 0.07051577867875927


: 

L'indice de silhouette sans ACP pour l'ensemble d'apprentissage est de 0.0567, ce qui indique une certaine superposition des clusters et une séparation moins claire entre eux.
L'indice de silhouette avec ACP pour l'ensemble d'apprentissage est de 0.0705, ce qui est légèrement plus élevé que celui sans ACP. Cela suggère que l'ACP a permis d'améliorer la séparation des clusters et donc la qualité de la classificat ion.
En conclusion, l'application de l'ACP à 90% de variance cumulée avant la classification a conduit à une amélioration de l'indice de silhouette pour l'ensemble d'apprentissage, indiquant une meilleure séparation des clusters et donc une meilleure performance du modèle de classification.